br/pkg/lightning/errormanager/errormanager.go

```diff
@@ -819,12 +749,84 @@ func (em *ErrorManager) ReplaceConflictKeys(
 			return errors.Trace(err)
 		}
 
+		// delete the additionally inserted rows for nonclustered PK
+		if err := exec.Transact(ctx, "delete additionally inserted rows for conflict detection 'replace' mode",
+			func(c context.Context, txn *sql.Tx) error {
+				sb := &strings.Builder{}
+				_, err2 := common.FprintfWithIdentifiers(sb, deleteNullDataRow, em.schema)
+				if err2 != nil {
+					return errors.Trace(err2)
+				}
+				_, err := txn.ExecContext(c, sb.String())
+				return errors.Trace(err)
+			}); err != nil {
+			return errors.Trace(err)
+		}
+
 		return nil
 	})
 
 	return errors.Trace(g.Wait())
 }
 
+// ResolveConflictKeysError query all conflicting rows (handle and their
+// values) from the current error report and return error
+// if the number of the conflicting rows is larger than 0.
+func (em *ErrorManager) ResolveConflictKeysError(
```

Comment:
it's better we implement the "error" setting in `dupeController.CollectRemoteDuplicateRows`. That function will fail earlier when see the first error. In this location we will record all errors which may cost a lot.

You can open another PR to do it.

Reply:
OK. Will implement in the next PR.